kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/expressions/ElementAtEvaluator.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (2023) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.defaults.internal.expressions;
+
+import java.util.Arrays;
+import static java.lang.String.format;
+
+import io.delta.kernel.data.ColumnVector;
+import io.delta.kernel.expressions.Expression;
+import io.delta.kernel.expressions.ScalarExpression;
+import io.delta.kernel.types.DataType;
+import io.delta.kernel.types.MapType;
+import io.delta.kernel.types.StringType;
+import io.delta.kernel.utils.Utils;
+
+import static io.delta.kernel.defaults.internal.DefaultKernelUtils.checkArgument;
+import static io.delta.kernel.defaults.internal.expressions.ImplicitCastExpression.canCastTo;
+
+/**
+ * Utility methods to evaluate {@code element_at} expression.
+ */
+class ElementAtEvaluator {
+    private ElementAtEvaluator() {}
+
+    /**
+     * Validate and transform the {@code element_at} expression with given validated and
+     * transformed inputs.
+     */
+    static ScalarExpression validateAndTransform(
+        ScalarExpression elementAt,
+        Expression mapInput,
+        DataType mapInputType,
+        Expression lookupKey,
+        DataType lookupKeyType) {
+
+        MapType asMapType = validateSupportedMapType(elementAt, mapInputType);
+        DataType keyTypeFromMapInput = asMapType.getKeyType();
+
+        if (!keyTypeFromMapInput.equivalent(lookupKeyType)) {
+            if (canCastTo(lookupKeyType, keyTypeFromMapInput)) {
+                lookupKey = new ImplicitCastExpression(lookupKey, keyTypeFromMapInput);
+            } else {
+                throw new UnsupportedOperationException(format(
+                    "%s: lookup key type (%s) is different from the map key type (%s)",
+                    elementAt, lookupKeyType, asMapType.getKeyType()));
+            }
+        }
+        return new ScalarExpression(elementAt.getName(), Arrays.asList(mapInput, lookupKey));
+    }
+
+    /**
+     * Utility method to evaluate the {@code element_at} on given map and key vectors.
+     * @param map {@link ColumnVector} of {@code map(string, string)} type.
+     * @param lookupKey {@link ColumnVector} of {@code string} type.
+     * @return
+     */
+    static ColumnVector eval(ColumnVector map, ColumnVector lookupKey) {
+        return new ColumnVector() {
+            // Store the last lookup value to avoid multiple looks up for same row id.
+            private int lastLookupRowId = -1;
```

Comment:
what's the intuition behind assuming this will be a common lookup case?

this is essentially just a cache of size 1. any reason not to have a cache of size, say, 10?